pkg/asset/storage.go

```diff
@@ -0,0 +1,369 @@
+		if err != nil {
+			return nil, fmt.Errorf("the call to MakeCloudStorageBackend failed: %w", err)
+		}
+	} else if strings.HasPrefix(cfg.UploadTo, "test://") {
+		backend = makeTestStorageBackend()
+	} else {
+		return nil, fmt.Errorf("unknown UploadTo during StorageFromConfig(): %#v", cfg.UploadTo)
+	}
+	return &Storage{
+		cfg:           cfg,
+		backend:       backend,
+		preprocessors: DefaultPreprocessors(),
+		dash:          dash,
+		tracer:        tracer,
+	}, nil
+}
+
+func (storage *Storage) AssetTypeEnabled(assetType Type) bool {
+	return storage.cfg.IsEnabled(assetType)
+}
+
+var ErrAssetTypeDisabled = errors.New("uploading assets of this type is disabled")
+
+func (storage *Storage) UploadFile(origFile string, assetType Type, name string) (string, error) {
+	storage.tracer.Log("UploadFile(%#v, %#v, %#v)", origFile, assetType, name)
+	if name == "" {
+		return "", fmt.Errorf("file name is not specified")
+	}
+	if !storage.AssetTypeEnabled(assetType) {
+		return "", fmt.Errorf("didn't upload an asset of type %s: %w",
+			assetType, ErrAssetTypeDisabled)
+	}
+	// The idea is to make a file name useful and yet unique.
+	// So we put a file to a pseudo-unique "folder".
+	path := fmt.Sprintf("%v/%s", time.Now().UnixNano(), name)
+	req := &uploadRequest{
+		origFile: origFile,
+		savePath: path,
+	}
+	handler := storage.preprocessors.GetPreprocessor(assetType)
+	res, err := handler.Do(req, func(req *uploadRequest) (*uploadResponse, error) {
+		return storage.backend.upload(req)
+	})
+	if err != nil {
+		return "", err
+	}
+	return res.downloadURL, nil
+}
+
+func (storage *Storage) UploadBuildAsset(origFile string, assetType Type, build *dashapi.Build) error {
+	baseName := filepath.Base(origFile)
+	fileExt := filepath.Ext(baseName)
+	name := fmt.Sprintf("%s-%s%s",
+		strings.TrimSuffix(baseName, fileExt),
+		build.KernelCommit,
+		fileExt)
+	url, err := storage.UploadFile(origFile, assetType, name)
+	if err != nil {
+		return err
+	}
+	// If the server denies the reques, we'll delete the orphaned file during deprecated files
+	// deletion later.
+	return storage.dash.AddBuildAsset(&dashapi.AddBuildAssetReq{
+		BuildID:     build.ID,
+		AssetType:   string(assetType),
+		DownloadURL: url,
+	})
+}
+
+func (storage *Storage) UploadBuildAssetStream(reader io.Reader, assetType Type, fileName string,
+	build *dashapi.Build) error {
+	dir, err := ioutil.TempDir("", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+	tmpFile := filepath.Join(dir, fileName)
+	w, err := os.Create(tmpFile)
+	if err != nil {
+		return fmt.Errorf("failed to create a temp file: %w", err)
+	}
+	_, err = io.Copy(w, reader)
+	if err != nil {
+		return fmt.Errorf("failed to write the reader stream: %w", err)
+	}
+	w.Close()
+	defer os.Remove(tmpFile)
+	return storage.UploadBuildAsset(tmpFile, assetType, build)
+}
+
+func (storage *Storage) UploadBuildAssetCopy(origFile string, assetType Type, copyName string,
```

Comment:
With 3 different versions UploadBuildAsset/UploadBuildAssetStream/UploadBuildAssetCopy it feels that the package it trying to be too helpful for callers. Only 2 of these versions are actually used. If also upload Crash, Bug and Manager artifacts, we will need 12 different permutations. What these variations add feels like osutil package territory. If we add something like osutil.WriteFileReader, it may be enough. Previously we avoided saving whole coverage report to disk and directly streamed it to GCS. Now we save it multiple times to disk anyway, so there is little point in providing special version for io.Reader.

Reply:
Hmm, it actually feels like moving everything to the streams might be the most generic option. Gzip compression/decompression is built on streams, xz can also operate on stdin/stdout (`echo "abd" | xz -0 -c | xxd -p`). So we can possibly get rid of all copying during internal processing.

But the problem is that sometimes we'd prefer to still copy a file before uploading it. E.g. for the case when uploading turned out to be very slow and the underlying file might have already been replaced by the newly compiled kernel. This is very unrealistic on syzbot, but we still need to somehow take that into account in the code.

Though that copying might be actually offloaded to `osutil` (e.g. `func CopyFileCallback(sourceFile string, cb func(tmpFile string)error) error`).

WDYT?